src/action/base/create_or_merge_nix_config.rs

```diff
@@ -0,0 +1,505 @@
+use nix::unistd::{chown, Group, User};
+use nix_config_parser::NixConfig;
+use tracing::{span, Span};
+
+use std::{
+    os::{unix::fs::MetadataExt, unix::fs::PermissionsExt},
+    path::{Path, PathBuf},
+};
+use tokio::{
+    fs::{remove_file, OpenOptions},
+    io::{AsyncSeekExt, AsyncWriteExt},
+};
+
+use crate::action::{Action, ActionDescription, ActionError, StatefulAction};
+
+const MERGEABLE_CONF_NAMES: &[&str] = &["experimental-features"];
+
+/** Create a file at the given location with the provided `buf`,
+optionally with an owning user, group, and mode.
+
+If `force` is set, the file will always be overwritten (and deleted)
+regardless of its presence prior to install.
+ */
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub struct CreateOrMergeNixConfig {
+    pub(crate) path: PathBuf,
+    user: Option<String>,
+    group: Option<String>,
+    mode: Option<u32>,
+    buf: String,
+    nix_configs: NixConfigs,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+struct NixConfigs {
+    pending_nix_config: NixConfig,
+    existing_nix_config: Option<NixConfig>,
+    merged_nix_config: Option<NixConfig>,
+}
+
+impl CreateOrMergeNixConfig {
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn plan(
+        path: impl AsRef<Path>,
+        user: impl Into<Option<String>>,
+        group: impl Into<Option<String>>,
+        mode: impl Into<Option<u32>>,
+        buf: String,
+    ) -> Result<StatefulAction<Self>, ActionError> {
+        let path = path.as_ref().to_path_buf();
+        let mode = mode.into();
+        let user = user.into();
+        let group = group.into();
+        // TODO: make path optional in nix_config_parser
+        let pending_nix_config =
+            nix_config_parser::parse_nix_config_string(buf.clone(), Path::new("/"))
+                .map_err(ActionError::ParseNixConfig)?;
+        let nix_configs = NixConfigs {
+            pending_nix_config,
+            existing_nix_config: None,
+            merged_nix_config: None,
+        };
+
+        let mut this = Self {
+            path,
+            user,
+            group,
+            mode,
+            buf,
+            nix_configs,
+        };
+
+        if this.path.exists() {
+            let metadata = this
+                .path
+                .metadata()
+                .map_err(|e| ActionError::GettingMetadata(this.path.clone(), e))?;
+
+            if let Some(mode) = mode {
```

Comment:
This is our third occurrence of this code so we should consider extracting it in a future PR....